README
NAME
    HTTP::ClientDetect - Detect language and location of an HTTP request

VERSION
    Version 0.02

SYNOPSIS
      use HTTP::ClientDetect::Location;
      my $geo = HTTP::ClientDetect::Location->new(db => $dbfile);
      my $country = $geo->request_country($request_object)

      use HTTP::ClientDetect::Language;
      my $lang_detect = HTTP::ClientDetect::Language->new(server_default => "hr_HR");
      my $lang = $lang_detect->language_short($request_object);

DESCRIPTION
    This module by itself doesn't do nothing. You have to load one or more
    of its children, as per synopsis.

    The object passed to the methods has to be an object which at least has
    an `header' method for the language detection, or `address' or
    `remote_address' for the country detection. This should work with
    Dancer::Request and Catalyst::Request objects.

AUTHOR
    Marco Pessotto, `<melmothx at gmail.com>'

BUGS
    Please report any bugs or feature requests to
    `bug-interchange6-plugin-autodetect at rt.cpan.org', or through the web
    interface at
    http://rt.cpan.org/NoAuth/ReportBug.html?Queue=HTTP-ClientDetect. I will
    be notified, and then you'll automatically be notified of progress on
    your bug as I make changes.

SUPPORT
    You can find documentation for this module with the perldoc command.

        perldoc HTTP::ClientDetect

    You can also look for information at:

    * RT: CPAN's request tracker (report bugs here)
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=HTTP-ClientDetect

    * AnnoCPAN: Annotated CPAN documentation
        http://annocpan.org/dist/HTTP-ClientDetect

    * CPAN Ratings
        http://cpanratings.perl.org/d/HTTP-ClientDetect

    * Search CPAN
        http://search.cpan.org/dist/HTTP-ClientDetect/

ACKNOWLEDGEMENTS
LICENSE AND COPYRIGHT
    Copyright 2013 Marco Pessotto.

    This program is free software; you can redistribute it and/or modify it
    under the terms of either: the GNU General Public License as published
    by the Free Software Foundation; or the Artistic License.

    See http://dev.perl.org/licenses/ for more information.



Changes
Revision history for HTTP-ClientDetect

0.02 Fri Jun 19 10:00:45 2015 CEST

	 * Initial release.



Makefile.PL
use 5.006;
use strict;
use warnings FATAL => 'all';
use ExtUtils::MakeMaker;

WriteMakefile(
    NAME             => 'HTTP::ClientDetect',
    AUTHOR           => q{Marco Pessotto <melmothx@gmail.com>},
    VERSION_FROM     => 'lib/HTTP/ClientDetect.pm',
    ABSTRACT_FROM    => 'lib/HTTP/ClientDetect.pm',
    LICENSE          => 'Perl_5',
    PL_FILES         => {},
    MIN_PERL_VERSION => 5.006,
    CONFIGURE_REQUIRES => {
        'ExtUtils::MakeMaker' => 0,
    },
    BUILD_REQUIRES => {
        'Test::More' => 0,
        'Moo' => 0,
        'Geo::IP::PurePerl' => 0, # but Geo::IP will do OK too
    },
    PREREQ_PM => {
        #'ABC'              => 1.6,
        #'Foo::Bar::Module' => 5.0401,
    },
    dist  => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
    clean => { FILES => 'HTTP-ClientDetect-*' },
);


lib/HTTP/ClientDetect.pm
package HTTP::ClientDetect;

use 5.006;
use strict;
use warnings FATAL => 'all';

=head1 NAME

HTTP::ClientDetect - Detect language and location of an HTTP request

=head1 VERSION

Version 0.02

=cut

our $VERSION = '0.02';


=head1 SYNOPSIS

  use HTTP::ClientDetect::Location;
  my $geo = HTTP::ClientDetect::Location->new(db => $dbfile);
  my $country = $geo->request_country($request_object)

  use HTTP::ClientDetect::Language;
  my $lang_detect = HTTP::ClientDetect::Language->new(server_default => "hr_HR");
  my $lang = $lang_detect->language_short($request_object);

=head1 DESCRIPTION

This module by itself doesn't do nothing. You have to load one or more
of its children, as per synopsis.

The object passed to the methods has to be an object which at least
has an C<header> method for the language detection, or C<address> or
C<remote_address> for the country detection. This should work with
L<Dancer::Request> and L<Catalyst::Request> objects.


=head1 AUTHOR

Marco Pessotto, C<< <melmothx at gmail.com> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-interchange6-plugin-autodetect at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=HTTP-ClientDetect>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.




=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc HTTP::ClientDetect


You can also look for information at:

=over 4

=item * RT: CPAN's request tracker (report bugs here)

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=HTTP-ClientDetect>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/HTTP-ClientDetect>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/HTTP-ClientDetect>

=item * Search CPAN

L<http://search.cpan.org/dist/HTTP-ClientDetect/>

=back


=head1 ACKNOWLEDGEMENTS


=head1 LICENSE AND COPYRIGHT

Copyright 2013 Marco Pessotto.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See L<http://dev.perl.org/licenses/> for more information.


=cut

1; # End of HTTP::ClientDetect


lib/HTTP/ClientDetect/Language.pm
package HTTP::ClientDetect::Language;

use 5.006;
use strict;
use warnings FATAL => 'all';
use Moo;

use Locale::Language;
use Locale::Country;

my @languages = all_language_codes();
my @countries = all_country_codes();

my %langs = map { $_ => 1 } @languages;
my %countrs  = map { $_ => 1 } @countries;

=head1 NAME

HTTP::ClientDetect::Language - Lookup the client's preferred language

=head1 VERSION

Version 0.01

=cut

our $VERSION = '0.01';


=head1 SYNOPSIS

    use HTTP::ClientDetect::Language;
    my $lang_detect = HTTP::ClientDetect::Language->new(server_default => "en_US");
    # inside a Dancer route
    get '/detect' => sub {
        my $req = request;
        my $lang = $lang_detect->language($req);
    }


=head1 ACCESSORS

=head2 server_default

The C<server_default> should be set in the constructor and defaults to
C<en_US>. This will be always returned if the lookup fails

=cut


has server_default => (is => 'rw',
                       default => sub { return "en_US" },
                       isa => sub {
                           die "Bad language $_[0]\n"
                             unless __PACKAGE__->check_language_name($_[0]);
                       });


=head2 available_languages

Accessor to an arrayref of languages available on the server side.
Please use the short version (C<de>, not C<de_DE>), otherwise the
check will be too restrictive.

=cut

has available_languages => (is => 'rw',
                            isa => sub {
                                my $aref = $_[0];
                                die "Not an arrayref" unless ref($aref) eq 'ARRAY';
                                foreach my $l (@$aref) {
                                    die "Bad language $l\n"
                                      unless __PACKAGE__->check_language_name($l);
                                }
                            },
                            default => sub { [] },
                           );


=head1 SUBROUTINES/METHODS

=head2 language($request_obj)

Return the preferred language of the request. The request object
should an object which has the methods C<accept_language> or C<header>

From L<http://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html>:

 The Accept-Language request-header field is similar to Accept, but
 restricts the set of natural languages that are preferred as a
 response to the request. Language tags are defined in section 3.10.

       Accept-Language = "Accept-Language" ":"
                         1#( language-range [ ";" "q" "=" qvalue ] )
       language-range  = ( ( 1*8ALPHA *( "-" 1*8ALPHA ) ) | "*" )

 Each language-range MAY be given an associated quality value which
 represents an estimate of the user's preference for the languages
 specified by that range. The quality value defaults to "q=1". For
 example,

       Accept-Language: da, en-gb;q=0.8, en;q=0.7

 would mean: "I prefer Danish, but will accept British English and
 other types of English." A language-range matches a language-tag if
 it exactly equals the tag, or if it exactly equals a prefix of the
 tag such that the first tag character following the prefix is "-".
 The special range "*", if present in the Accept-Language field,
 matches every tag not matched by any other range present in the
 Accept-Language field.

      Note: This use of a prefix matching rule does not imply that
      language tags are assigned to languages in such a way that it is
      always true that if a user understands a language with a certain
      tag, then this user will also understand all languages with tags
      for which this tag is a prefix. The prefix rule simply allows the
      use of prefix tags if this is the case.

 The language quality factor assigned to a language-tag by the
 Accept-Language field is the quality value of the longest language-
 range in the field that matches the language-tag. If no language-
 range in the field matches the tag, the language quality factor
 assigned is 0. If no Accept-Language header is present in the
 request, the server

 SHOULD assume that all languages are equally acceptable. If an
 Accept-Language header is present, then all languages which are
 assigned a quality factor greater than 0 are acceptable.

  It might be contrary to the privacy expectations of the user to send
  an Accept-Language header with the complete linguistic preferences
  of the user in every request

=cut

sub language {
    my ($self, $obj) = @_;
    my @browser_langs = $self->browser_languages($obj);
    my @avail = @{$self->available_languages};
    if (@avail) {
        foreach my $ua_lang (@browser_langs) {
            foreach my $avail_lang (@avail) {
                if ($ua_lang =~ m/^\Q$avail_lang\E(_[A-Z]+)?$/) {
                    return $ua_lang;
                }
            }
        }
        # nothing? then return the server default
        return $self->server_default;
    }
    else {
        return $browser_langs[0];
    }
}

=head2 browser_languages($request)

This method returns the parsed and sorted list of language preferences
set in the browser, when the first element has higher priority.

=cut

sub browser_languages {
    my ($self, $obj) = @_;
    return $self->server_default unless $obj;
    my $accept_str;
    if ($obj->can("accept_language")) {
        $accept_str = $obj->accept_language;
    }
    # nothing? try with header, but don't count too much on this
    if (!$accept_str and $obj->can("header")) {
        $accept_str = $obj->header('Accept-Language');
    }
    return $self->server_default unless $accept_str;
    
    # split the string at ,
    my @langs = split(/\s*,\s*/, $accept_str);
    my @to_order;
    foreach my $lang_str (@langs) {
        next unless $lang_str;
        my ($q, $code);
        if ($lang_str =~ m/([a-zA-Z]+([-_][a-zA-Z]+)?)\s*(;\s*q\s*=\s*([0-9\.]+))?/) {
            $code = $self->check_language_name($1);
            $q = $4 || 1;
        }
        next unless $code;
        push @to_order, [ $code => $q ];
        # sort by q
    }
    return $self->server_default unless @to_order;
    my @ordered = sort { $b->[1] <=> $a->[1] } @to_order;
    return map { $_->[0] } @ordered;
}

=head3 language_short($request_obj)

Return the short language version (i.e.), the language name without
the country part.

=cut

sub language_short {
    my ($self, $obj) = @_;
    my $lang = $self->language($obj);
    # strip the second part
    $lang =~ s/_.*$//;
    return $lang;
}


=head3 check_language_name

Returns a normalized version of the language name, lower case for the
language, upper case for the country. Undef it was not possible to
validate it.

=cut

sub check_language_name {
    my ($self, $code) = @_;
    my ($lang, $country);
    return unless $code;
    if ($code =~ m/([a-zA-Z]{2})([_-]([a-zA-Z]*))?/) {
        $lang = $1;
        $country = $3 || "";
    }
    else {
        $lang = $code;
        $country = $code; # eg. fr fr
    }
    # lowercase;
    $lang = lc($lang);
    $country = lc($country);
    # check the lang;
    return unless $langs{$lang};
    # if the country doesn't validate, we fix the common scenario (en
    # => US), and append the same
    if ($countrs{$country}) {
        return $lang . "_" . uc($country);
    }
    # then do some heuristics, if the country didn't match
    if ($lang eq 'en') {
        return $lang . "_US";
    }
    # then try the language as a country
    if ($countrs{$lang}) {
        return $lang . "_" . uc($lang);
    }
    # if we are still here, return the language, there are cases we
    # can't catch, like ja_JP
    return $lang;
}


=head1 AUTHOR

Marco Pessotto, C<< <melmothx at gmail.com> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-interchange6-plugin-autodetect at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=HTTP-ClientDetect>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.




=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc HTTP::ClientDetect::Language


You can also look for information at:

=over 4

=item * RT: CPAN's request tracker (report bugs here)

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=HTTP-ClientDetect>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/HTTP-ClientDetect>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/HTTP-ClientDetect>

=item * Search CPAN

L<http://search.cpan.org/dist/HTTP-ClientDetect/>

=back


=head1 ACKNOWLEDGEMENTS


=head1 LICENSE AND COPYRIGHT

Copyright 2013 Marco Pessotto.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See L<http://dev.perl.org/licenses/> for more information.


=cut

1; # End of HTTP::ClientDetect::Language


lib/Interchange6/Plugin/Autodetect.pm
package Interchange6::Plugin::Autodetect;

use 5.006;
use strict;
use warnings FATAL => 'all';

=head1 NAME

Interchange6::Plugin::Autodetect - The great new Interchange6::Plugin::Autodetect!

=head1 VERSION

Version 0.02

=cut

our $VERSION = '0.02';


=head1 SYNOPSIS

Quick summary of what the module does.

Perhaps a little code snippet.

    use Interchange6::Plugin::Autodetect;

    my $foo = Interchange6::Plugin::Autodetect->new();
    ...

=head1 EXPORT

A list of functions that can be exported.  You can delete this section
if you don't export anything, such as for a purely object-oriented module.

=head1 SUBROUTINES/METHODS

=head2 function1

=cut

sub function1 {
}

=head2 function2

=cut

sub function2 {
}

=head1 AUTHOR

Marco Pessotto, C<< <melmothx at gmail.com> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-interchange6-plugin-autodetect at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Interchange6-Plugin-Autodetect>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.




=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Interchange6::Plugin::Autodetect


You can also look for information at:

=over 4

=item * RT: CPAN's request tracker (report bugs here)

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Interchange6-Plugin-Autodetect>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Interchange6-Plugin-Autodetect>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Interchange6-Plugin-Autodetect>

=item * Search CPAN

L<http://search.cpan.org/dist/Interchange6-Plugin-Autodetect/>

=back


=head1 ACKNOWLEDGEMENTS


=head1 LICENSE AND COPYRIGHT

Copyright 2013 Marco Pessotto.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See L<http://dev.perl.org/licenses/> for more information.


=cut

1; # End of Interchange6::Plugin::Autodetect


lib/Interchange6/Plugin/Interchange5/Request.pm
package Interchange6::Plugin::Interchange5::Request;

use 5.006;
use strict;
use warnings FATAL => 'all';
use Moo;

=head1 NAME

Interchange6::Plugin::Interchange5::Request - Mimic Dancer::Request inside IC5

=head1 VERSION

Version 0.01

=cut

our $VERSION = '0.01';


=head1 SYNOPSIS

Quick summary of what the module does.

In a tag (shipped as dancer_request.tag)

    use Interchange6::Plugin::Interchange5::Request;
    my %env = %{::http()->{env}};
    return Interchange6::Plugin::Interchange5::Request->new(env => \%env);
    ...

Somewhere else

    my $req = $Tag->dancer_request;
    $req->header('Accept-Language');
    $req->accept_language;
    ....


=head1 ACCESSORS

=head2 env

=cut

has env => (is => 'ro',
            required => 1,
            isa => sub {
                die unless ref($_[0]) eq 'HASH';
            });

=head1 METHODS

=head2 environment($name)

Look into the environment variables, with the following routine:
first, we uppercase the name and replace any non-alpha and non-digit
character with the underscore. Then we look into the environment. If
not found, we try to prepend HTTP_. Return undef in nothing is found.

=head2 header($name)

Alias for environment

=cut

sub header {
    my ($self, $name) = @_;
    return $self->environment($name);
}

sub environment {
    my ($self, $name) = @_;
    return unless $name;
    $name =~ s/[^a-zA-Z0-9_]/_/g;
    $name = uc($name);
    my $http_name = "HTTP_" . $name;
    my $env = $self->env;
    if (exists $env->{$name}) {
        return $env->{$name};
    }
    elsif (exists $env->{$http_name}) {
        return $env->{$http_name};
    }
    else {
        return;
    }
}


=head2 SHORTCUTS

The following methods are just shortcuts for the C<environment> method.

=over 4

=item  accept

=item  accept_charset

=item  accept_encoding

=item  accept_language

=item  accept_type

=item  agent (alias for "user_agent")

=item  connection

=item  forwarded_for_address

=item  forwarded_protocol

=item  forwarded_host

=item  host

=item  keep_alive

=item  path_info

=item  referer

=item  remote_address

=item  user_agent

=back

=cut

sub accept { return shift->environment("accept") }

sub accept_charset { return shift->environment("accept_charset") }

sub accept_encoding { return shift->environment("accept_encoding") }

sub accept_language { return shift->environment("accept_language") }

sub accept_type { return shift->environment("accept_type") }

sub agent { return shift->environment("user_agent") }

sub connection { return shift->environment("connection") }

sub forwarded_for_address { return shift->environment("forwarded_for_address") }

sub forwarded_protocol { return shift->environment("forwarded_protocol") }

sub forwarded_host { return shift->environment("forwarded_host") }

sub host { return shift->environment("host") }

sub keep_alive { return shift->environment("keep_alive") }

sub path_info { return shift->environment("path_info") }

sub referer { return shift->environment("referer") }

sub remote_address { return shift->environment("remote_addr") }

sub user_agent { return shift->environment("user_agent") }

=head1 AUTHOR

Marco Pessotto, C<< <melmothx at gmail.com> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-interchange6-plugin-autodetect at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Interchange6-Plugin-Autodetect>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.




=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Interchange6::Plugin::Interchange5::Request


You can also look for information at:

=over 4

=item * RT: CPAN's request tracker (report bugs here)

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Interchange6-Plugin-Autodetect>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Interchange6-Plugin-Autodetect>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Interchange6-Plugin-Autodetect>

=item * Search CPAN

L<http://search.cpan.org/dist/Interchange6-Plugin-Autodetect/>

=back


=head1 ACKNOWLEDGEMENTS


=head1 LICENSE AND COPYRIGHT

Copyright 2013 Marco Pessotto.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See L<http://dev.perl.org/licenses/> for more information.


=cut

1; # End of Interchange6::Plugin::Interchange5::Request


lib/HTTP/ClientDetect/Location.pm
package HTTP::ClientDetect::Location;

use 5.006;
use strict;
use warnings FATAL => 'all';
use Module::Load;
use Moo;

=head1 NAME

HTTP::ClientDetect::Location - Lookup the country of the client using Geo::IP

=head1 VERSION

Version 0.01

=cut

our $VERSION = '0.01';


=head1 SYNOPSIS

    use HTTP::ClientDetect::Location;
    my $geo = HTTP::ClientDetect::Location->new(db => "/path/to/geo-ip");
    # inside a Dancer route
    get '/detect' => sub {
        my $req = request;
        my $country_code = $geo->country_code($req);
    }


=head1 ACCESSORS

=head2 db

Path to the geoip database, which can be retrieved from
http://dev.maxmind.com/geoip/legacy/geolite/

C<db> must be passed to the constructor.

=cut

has db => (is => 'ro',
           required => 1,
           isa => sub {
               die "db is not a file" unless -f $_[0];
           });


=head2 geo

This accessor wraps L<Geo::IP> or L<Geo::IP::PurePerl>. Calling
$geo->geo will return an instantiated object, and you can call, e.g.
C<$geo->geo->country_code_by_addr("128.31.0.51")> or
C<$geo->geo->country_code_by_name("linuxia.de")>.

=cut

has geo => (is => 'rwp',
            builder => 1);

sub _build_geo {
    my $self = shift;
    my $module = "Geo::IP";
    eval {
        load $module;
    };
    if ($@) {
        $module .= "::PurePerl";
        # if we die here too bad
        load $module;
    }
    my $gi = $module->open($self->db);
    return $gi;
}

=head1 SUBROUTINES/METHODS

=head2 request_country($request_obj)

Return the country of the request. The request should be a string with
the ip or L<Dancer::Request> object (or something that respond to the
method "remote_address").

=cut

sub request_country {
    my ($self, $arg) = @_;
    my $ip;
    if (ref($arg)) {
        # Dancer
        if ($arg->can("remote_address")) {
            $ip = $arg->remote_address;
        }
        # Catalyst
        elsif ($arg->can("address")) {
            $ip = $arg->address;
        }
    }
    else {
        $ip = $arg;
    }
    return unless $ip;
    if ($ip =~ m/(([0-9]{1,3}\.){3}([0-9]{1,3}))/) {
        $ip = $1;
        return $self->geo->country_code_by_addr($ip);
    }
    else {
        return;
    }
}

=head1 AUTHOR

Marco Pessotto, C<< <melmothx at gmail.com> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-interchange6-plugin-autodetect at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=HTTP-ClientDetect>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.




=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc HTTP::ClientDetect::Location


You can also look for information at:

=over 4

=item * RT: CPAN's request tracker (report bugs here)

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=HTTP-ClientDetect>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/HTTP-ClientDetect>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/HTTP-ClientDetect>

=item * Search CPAN

L<http://search.cpan.org/dist/HTTP-ClientDetect/>

=back


=head1 ACKNOWLEDGEMENTS


=head1 LICENSE AND COPYRIGHT

Copyright 2013 Marco Pessotto.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See L<http://dev.perl.org/licenses/> for more information.


=cut

1; # End of HTTP::ClientDetect::Location
